serverless/log-email-sns/src/query.ts

```diff
@@ -0,0 +1,145 @@
+ */
+export const updateMessageWithSuccess = async (
+  metadata: Metadata
+): Promise<number | undefined> => {
+  console.log(`Updating email_messages table. ${JSON.stringify(metadata)}`)
+  // Since notifications for the same messageId can be interleaved, we only update that message if this notification is newer than the previous.
+  const [result] =
+    (await sequelize?.query(
+      `UPDATE email_messages SET received_at=:timestamp, updated_at = clock_timestamp(), status='SUCCESS' 
+      WHERE id=:id 
+      AND (received_at IS NULL OR received_at < :timestamp)
+      RETURNING campaign_id;`,
+      {
+        replacements: { timestamp: metadata.timestamp, id: metadata.id },
+        type: QueryTypes.UPDATE,
+      }
+    )) || []
+
+  const campaign: { campaign_id: number } = (result || [])[0]
+  return campaign?.campaign_id
+}
+
+export const haltCampaignIfThresholdExceeded = async (campaignId?: number) => {
+  if (campaignId === undefined) {
+    return
+  }
+  // Compute threshold
+  const [result] =
+    (await sequelize?.query(
+      `SELECT SUM(CASE WHEN status='INVALID_RECIPIENT' THEN 1 ELSE 0 END) AS invalid, COUNT(1) AS running_total FROM email_messages WHERE campaign_id=:campaignId AND status IS NOT NULL`,
+      {
+        replacements: { campaignId },
+        type: QueryTypes.SELECT,
+      }
+    )) || []
+
+  const {
+    invalid,
+    running_total,
+  }: { invalid?: number; running_total?: number } = result
+
+  if (invalid !== undefined && running_total !== undefined) {
+    const percentageInvalid = invalid / running_total
+    const exceedsHaltNumber = invalid > config.get('minHaltNumber')
+    const exceedsHaltPercentage =
+      percentageInvalid > config.get('minHaltPercentage')
+    console.log(
+      `Current campaign_id=${campaignId} invalid=${invalid} running_total=${running_total} 
+      percentageInvalid=${percentageInvalid} 
+      config.minHaltNumber=${config.get('minHaltNumber')}
+      config.minHaltPercentage=${config.get('minHaltPercentage')}
+      exceedsHaltNumber=${exceedsHaltNumber}
+      exceedsHaltPercentage=${exceedsHaltPercentage}`
+    )
+    /* With default MIN_HALT_NUMBER=10, MIN_HALT_PERCENTAGE=0.1, 
+    it means that 
+    - if there were 11 messages sent thus far, and 11 invalid recipients, the campaign would halt immediately since 11/11 > MIN_HALT_PERCENTAGE
+    - if there were 110 messages sent thus far, and 11 invalid recipients, the campaign would not halt, since 11/110 <= MIN_HALT_PERCENTAGE
+    */
+    if (exceedsHaltNumber && exceedsHaltPercentage) {
+      // Halt
+      console.log(
+        `Halting campaign_id=${campaignId} invalid=${invalid} running_total=${running_total} percentageInvalid=${percentageInvalid}`
+      )
+
+      try {
+        await sequelize?.transaction(async (_t) => {
+          await sequelize?.query(`SELECT stop_jobs(:campaignId)`, {
+            replacements: { campaignId },
+            type: QueryTypes.SELECT,
+          })
```

Comment:
why is the transaction not used in the param?

Reply:
you are right